addons/l10n_id/tests/test_qris.py

```diff
@@ -0,0 +1,193 @@
+                "qris_invoiceid": "413255111",
+                "qris_nmid": "ID1020021181745"
+            }
+        }
+        cls.success_qris_get_second = {
+            "status": "success",
+            "data": {
+                "qris_content": "Test Content Second",
+                "qris_request_date": "2024-02-27 11:13:42",
+                "qris_invoiceid": "413255111",
+                "qris_nmid": "ID1020021181745"
+            }
+        }
+
+        cls.qris_status_fail = {
+            "status": "failed",
+            "data": {
+                "qris_status": "unpaid"
+            }
+        }
+
+        cls.qris_status_success = {
+            "status": "success",
+            "data": {
+                "qris_status": "paid",
+                "qris_payment_customername": "Zainal Arief",
+                "qris_payment_methodby": "Sakuku"
+            },
+            "qris_api_version_code": "2206091709"
+        }
+
+    @freeze_time("2024-02-27 04:15:00")
+    def test_qris_qr_code_generation(self):
+        """ QR-Code generation conditions:
+
+        - should only come from portal side
+        - only call the API when the QRIS transaction not found or lasted for > 30 minutes
+        - if transaction found and < 30 minutes, no API call needed
+        """
+        self.qris_qr_invoice.qr_code_method = 'id_qr'
+
+        with patch(
+            'odoo.addons.l10n_id.models.res_bank.ResBank._make_qris_request', return_value=self.success_qris_get
+        ) as patched:
+            # QR code shouldn't be generated without the context.
+            result = self.qris_qr_invoice._generate_qr_code()
+            self.assertIsNone(result)
+
+            # But of course, should be with it.
+            result = self.qris_qr_invoice.with_context({'is_online_qr': True})._generate_qr_code()
+            self.assertIsNotNone(result)
+
+            # Confirm that the QR was successfully registered on the invoice.
+            qr_details = self.qris_qr_invoice.qris_invoice_details
+            self.assertIsNotNone(qr_details)
+
+            qris_data = self.success_qris_get['data']
+            self.assertEqual(qr_details[0]['qris_invoice_id'], qris_data['qris_invoiceid'])
+            self.assertEqual(self.qris_qr_invoice.latest_qris_content, qris_data['qris_content'])
+            patched.assert_called_once()
+
+            # Check the API is not called again, as it should reuse the existing QR until it is expired
+            self.qris_qr_invoice.with_context({'is_online_qr': True})._generate_qr_code()
+            patched.assert_called_once()
+
+            # Check that if the qr code has expired, we correctly generate a new one.
+            qr_details[0]['qris_creation_datetime'] = '2024-02-27 03:00:00'
+            self.qris_qr_invoice.qris_invoice_details = qr_details
+            self.qris_qr_invoice.with_context({'is_online_qr': True})._generate_qr_code()
+            self.assertEqual(patched.call_count, 2)
```

Comment:
Since you include the log note logic, maybe also add that checking as part of unit test ( like the followinghttps://github.com/odoo-dev/odoo/blob/ebf255d8f44fec80bbbc339ca9077a094bc125a7/addons/sale/tests/test_sale_order.py#L123)

Reply:
Added a check for that in the test_fetch_payment_status_success test

Reply:
looks good!